dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/batching/DefaultConsentAwareDataHandler.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.data.batching
+
+import com.datadog.android.core.internal.data.privacy.Consent
+import com.datadog.android.core.internal.data.privacy.ConsentProvider
+import com.datadog.android.core.internal.data.privacy.ConsentProviderCallback
+
+internal class DefaultConsentAwareDataHandler<T>(
+
+    private val consentProvider: ConsentProvider,
+    private val intermediaryDataProcessor: DataProcessor<T>,
+    private val targetDataProcessor: DataProcessor<T>,
+    private val batchedDataMigrator: BatchedDataMigrator
```

Comment:
IMO there should be only one `ConsentAwareDataHandler<T>` using single `processor: DataProcessor<T>` interface and swapping the processor implementations on consent change. The migrator should be decoupled from here.

There are two distinct things to consider:
* consent **value** - decides if any should be persisted at all,
* consent **transition** - decides on how to move data around.

The single `processor: DataProcessor<T>` could be either:
* `UnsetConsentProcessor` - writes all incoming data to intermediate folder,
* `YesConsentProcessor` - writes all incoming data to upload directory,
* `NoConsentProcessor` - drops all incoming data.

The middle one will have a specialised variant for RUM:
* `RUMYesConsentProcessor` - writes all incoming data to `/authorised` folder and acts on RUM session completions.

The handler gets initialised with 3 factory functions: for `unset` / `yes` and `no` consent values. This way we keep one handler which can manage all 3 features and its only responsibilities are: swapping + ensuring thread safety.

When consent value is changed, the handler swaps to the new processor basing on the **consent's new value**. It uses the **consent transition** (so new + old value) to decide on which `migrator` to pass to the processor. There could be as many migrators as we have significant transitions + specialised variants for RUM:
* `UnsetToYesDataMigrator` - moves data from `/intermediate` to `/ready-for-upload` folder,
* `UnsetToNoDataMigrator`- deletes all data in `/intermediate` folder,
* `YesToUnsetDataMigrator`- deletes all data in `/ready-for-upload` folder,
* ...
* `RUMYesToUnsetDataMigrator`- deletes all data in `/ready-for-upload` and `/authorised` folders,
* ...

The processor executes the `migrator` on its initialisation, which happens during in a thread-safe manner during the swap in `handler`. The only responsibility of migrator is to move data around.

All the complexity of RUM session sampling is only impacting the `RUMYesConsentProcessor`, and IMO there's no need to manage it in on the consent handler level.